### PR TITLE
chore(vrl, schemas): add new `value` crate, including `Kind` type

### PR DESCRIPTION
**Pull request by JeanMertz**
This commit adds a new `value` crate, which — to start — contains the `Kind` type.

This type is inspired by VRL's `TypeDef` kind, but:

- Is completely separate from VRL, or it's `Value` type.
- Has an improved/safer/easier-to-use API.
- Does not include the concept of "fallibility".
- Is properly unit-tested.

The goal of this crate is to share major components of VRL's progressive type checking between VRL and Vector, as Vector also learns about even field types, as part of the ongoing "schema" work.

This PR does not yet integrate the crate into either VRL or Vector. That work will happen in follow-up PRs, to prevent this PR from becoming too big, and including lots of mechanical changes, making it harder to review.

### CODE REVIEW

File: lib/value/src/kind.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+/// this library to expose.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Kind {
+    bytes: Option<()>,
+    integer: Option<()>,
+    float: Option<()>,
+    boolean: Option<()>,
+    timestamp: Option<()>,
+    regex: Option<()>,
+    null: Option<()>,
```
Comment (JeanMertz) [inline lib/value/src/kind.rs:33]: Just a note that I started with booleans, but preferred having all fields be `Option`s, since the collection fields needed that anyway. These fields aren't public, but it made for a more consistent internal API.

Also, `size_of::<bool>() == size_of::<Option<()>>()`, so there's no performance loss (not that it matters too much for this type and the way it is used).

File: lib/value/src/kind/accessor.rs [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+                    .as_ref()
+                    .and_then(|collection| collection.known().get(&(field.into())))?,
+
+                // Try finding one of the fields in the existing object.
+                Segment::Coalesce(fields) => kind.object.as_ref().and_then(|collection| {
+                    let field = fields
+                        .iter()
+                        .find(|field| collection.known().contains_key(&((*field).into())))?;
+
+                    collection.known().get(&(field.into()))
```
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:36]: I'm not sure if this is correct.

Say we have an object `{thing: int | null, thung: float}` and we do `find_at_path(".(thing | thung)")`. 

I believe this should return `int | float`, since the resulting type at runtime of fetching that path would either be an `int` or a `float`, depending on the value of `thing` at runtime.

This function would currently return `int | null`.

The coalesce needs to fold the types of all the potentially nullable fields together.

File: lib/value/src/kind/accessor.rs [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+            match segment {
+                Segment::Field(field) => {
+                    self = object_from_field(field, self);
+                }
+                Segment::Coalesce(fields) => {
+                    // We pick the last field in the list of coalesced fields, there is no
+                    // "correct" way to handle this case, other than not supporting it, or making
+                    // this method call fallible.
+                    let field = fields.last().expect("at least one");
+                    self = object_from_field(field, self);
```
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:72]: Couldn't we add the kind to all fields indicated in the coalesce?

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:72]: We could, but I wouldn't actually call that "correct" either.

I consider coalesced paths to be "start from the first, and progress until one matches", and when inserting, I consider it to be "take the first, no matter what" (which I realize is different from this, so I should update this).

File: lib/value/src/kind/accessor.rs (hunk 2 above)
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:36]: Good catch!

File: lib/value/src/kind/comparison.rs [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        let h = self.is_array();
+        if !(!h || !a && !b && !c && !d && !e && !f && !g) {
+            return false;
+        }
+
+        let i = self.is_object();
+        if !(!i || !a && !b && !c && !d && !e && !f && !g && !h) {
+            return false;
+        }
+
```
Comment (StephenWakely) [inline lib/value/src/kind/comparison.rs:175]: I think a slightly easier to follow, albeit marginally slower implementation would be:

```rust
let mut count = 0;
if self.is_bytes() {
  count += 1;
}
if self.is_integer(){
  count += 1;
}
if self.is_float(){
  count += 1;
}
if self.is_boolean(){
  count += 1;
}
if self.is_timestamp(){
  count += 1;
}
if self.is_regex() {
  count += 1;
}
if self.is_null(){
  count += 1;
}
if self.is_array(){
  count += 1;
}
if self.is_object(){
  count += 1;
}
count == 1
```

File: lib/value/src/kind/accessor.rs [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+                    })
+                }
+
+                // Try finding one of the fields in the existing object.
+                Segment::Coalesce(fields) => kind.object.as_mut().and_then(|collection| {
+                    let field = fields
+                        .iter()
+                        .find(|field| collection.known().contains_key(&((*field).into())))?;
+
+                    collection.known_mut().get_mut(&(field.into()))
```
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:152]: I think this should recurse down all matching fields in the coalesce to update all matching paths.

**Review by StephenWakely — DISMISSED** [review]
A few issues with the coalesce functionality, which is unfortunate since they are it seems hardly ever used, but worth getting right if we decide to continue supporting them!

File: lib/value/src/kind/accessor.rs (hunk 5 above)
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:152]: Similar to above, I don't think it should. It should stop at the first matched field, similar to how querying a coalesced path stops at querying the first matched field.

File: lib/value/src/kind/comparison.rs (hunk 4 above)
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:175]: Agreed. That is easier to reason about, and performance isn't that much of an issue in this code path.

File: lib/value/src/kind/accessor.rs (hunk 5 above)
Comment (StephenWakely) [reply lib/value/src/kind/accessor.rs:152]: Given this typedef, `{thing: { thang: int } | null, thung: { thang: int } }`.

The typedef for that could represent both objects:

`{ thing: { thang: 32 }, thung: { thang: 43 } }`
`{ thung: { thang: 43 } }`

Calling `del(x, .(thing | thung).thang)` would result in:

`{ thing: {}, thung: { thang: 43 } }`
`{ thung: {} }`

Thereby the typedef for the object should represent the fact that `thang` can be removed from both branches.

If you call `remove_at_path(.(thing | thung).thang)`, currently the result would be: `{ thung: { thang: float }`, implying that `thang` can never be removed from `thung`.

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:152]: Ah yes, you are totally correct. Ugh, these coalesced paths are such a mess to work with. I'm curious if we actually handled this correctly before. Regardless, it should work as expected, so I'll look into this.

Depending on the complexity, I might do it in a follow-up PR so that I'm not blocked on moving the schema work forward for now. It _might_ be the case that I can no longer take a `&mut self` but have to take the `Kind` by value and return a cloned variant, which would be unfortunate, but the price of getting this right. Although it might still be possible to keep the mutable reference, I'll have to see how much the borrow-checker complains (implementing these functions using `&mut self` was quite a bit of work to get right, the old implementation used cloning to work around that).

Comment (StephenWakely) [reply lib/value/src/kind/accessor.rs:152]: I wonder if we could just remove all type defs at the first coalesce and say anything goes after that?

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:152]: That would be a "correct, but non-complete" way to tackle this. I think I'm for this approach to start, we can refine it if we really see a need to do so.

File: lib/value/src/kind/comparison.rs [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+            return true;
+        }
+
+        if self.is_array() && other.is_array() {
+            return true;
+        }
+
+        if self.is_object() && other.is_object() {
+            return true;
+        }
```
Comment (pablosichert) [inline lib/value/src/kind/comparison.rs:270]: Where / how do we use the `intersects` method? I'd like to have some context to reason about the collection cases.

**Review by pablosichert — COMMENTED** [review]
Super nice work, @JeanMertz! That was quite some involved logic to unpack. The overall state looks really good to me. There's just a couple of questions I have to clarify my understanding to give a final approval.

File: lib/value/src/kind/comparison.rs (hunk 4 above)
Comment (pablosichert) [reply lib/value/src/kind/comparison.rs:175]: (Might also just use `self.is_bytes() + ... + self.is_object() == 1`.)

File: lib/value/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "value"
+version = "0.1.0"
+authors = ["Vector Contributors <vector@timber.io>"]
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/Cargo.toml:4]: Ah yeah, I mostly keep copying this from one VRL lib to another. Will get the correct metadata before I merge this 👍

File: lib/value/src/kind/comparison.rs [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+        self.object.is_some()
+    }
+
+    /// Returns `true` if exactly one type is set.
+    ///
+    /// For example, the following:
+    ///
+    /// ```rust,ignore
+    /// kind.is_float() && kind.is_exact()
+    /// ```
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:129]: Yeah, that's a nice addition. I briefly contemplated adding doc-tests everywhere, but didn't go that route just yet. I'll see if I can add a few here and there, it's straightforward to do, just takes a bunch of time to get them all documented.

File: lib/value/src/kind/accessor.rs [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+            (
+                "array w/o matching path",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[2]").unwrap(),
+                    want: None,
+                },
+            ),
+            (
+                "array w/ matching path, shifting indices",
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:269]: I suspect I copied this example from the `remote_at_path` tests, since they are common in structure, just with different results. This one doesn't make sense for `find_at_path` though, since no mutation happens of the original event, will update 👍

File: lib/value/src/kind/accessor.rs [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+                    want: Kind::empty().or_array(BTreeMap::from([(2.into(), Kind::integer())])),
+                },
+            ),
+            (
+                "array negative index",
+                TestCase {
+                    kind: Kind::integer(),
+                    path: LookupBuf::from_str("[-2]").unwrap(),
+                    want: Kind::empty().or_array(BTreeMap::default()),
+                },
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:370]: Yeah, this behaviour is questionable. I'll think on this, and will circle back to this in the follow-up PR that links it up to VRL, if I think there's a better way to handle this (preferably without panicking, and most-likely also without returning an error because there are places in VRL where this method is called that doesn't allow returning an error. Although I might opt to return an error for this method, but then panic at the VRL call-site, if we deem it to be a programming error).

File: lib/value/src/kind/accessor.rs [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+            ),
+            (
+                "array w/ negative indexing",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[-1]").unwrap(),
+                    returned: None,
+                    mutated: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                },
+            ),
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:478]: It's not a programming error per sé, there are places where people can use negative indexing in paths. But yeah, it is an edge-case, and one we (currently) don't really support.

I can see an update to the type system in the future that would allow us to better track array sizes, and then allowing or rejecting negative indexing more explicitly.

File: lib/value/src/kind/builder.rs [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+
+// `remove_*` methods to narrow the state of a type.
+impl Kind {
+    /// Remove the `bytes` state from the type.
+    ///
+    /// If the type already excluded this state, the function returns `Ok(false)`.
+    ///
+    /// # Errors
+    ///
+    /// If removing this state leaves an "empty" type, then the error variant is returned. This was
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/builder.rs:375]: Good suggestion, will update the rest as well.

File: lib/value/src/kind/collection.rs [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+    #[must_use]
+    pub(super) fn from_parts(known: BTreeMap<T, Kind>, unknown: impl Into<Unknown>) -> Self {
+        Self {
+            known,
+            unknown: unknown.into(),
+        }
+    }
+
+    /// Create a new collection with a defined "unknown fields" value, and no known fields.
+    pub fn unknown(unknown: impl Into<Unknown>) -> Self {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:41]: Interesting. Not sure, I don't think Clippy warned me of this one, but I might have overlooked the error. I'll take a peek before merging.

File: lib/value/src/kind/collection.rs [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+            unknown: Unknown::json(),
+        }
+    }
+
+    /// Check if the collection fields can be of any kind.
+    ///
+    /// This returns `false` if at least _one_ field kind is known.
+    #[must_use]
+    pub fn is_any(&self) -> bool {
+        self.known.iter().all(|(_, k)| k.is_any()) && self.unknown.is_any()
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:71]: It should, good suggestion.

File: lib/value/src/kind/collection.rs [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+    }
+
+    /// Get the "known" field value kinds.
+    #[must_use]
+    pub fn known(&self) -> &BTreeMap<T, Kind> {
+        &self.known
+    }
+
+    /// Get a mutable reference to "known" field value kinds.
+    pub fn known_mut(&mut self) -> &mut BTreeMap<T, Kind> {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:87]: I think Clippy doesn't warn here because we take `&mut self`, which often (but not always, such as here) implies the method has side effects, that don't always require the use of the return value.

File: lib/value/src/kind/collection.rs [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+    }
+
+    /// Merge the `other` collection into `self`.
+    ///
+    /// The following merge strategies are applied.
+    ///
+    /// For *known fields*:
+    ///
+    /// - If a field exists in both collections, their `Kind`s are merged, or the `other` fields
+    ///   are used (depending on the configured [`Strategy`](merge::Strategy).
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:149]: It's mostly to be consistent with the existing implementation (i.e. to make all `stdlib` tests pass).

We'll probably want to come up with a better name for the strategy, but I'm not sure what to call it. I'll think about this a bit when I do another pass once the library is integrated into VRL.

File: lib/value/src/kind/collection.rs [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+                    other: Collection::unknown(Kind::bytes()),
+                    want: false,
+                },
+            ),
+            (
+                "other-known match",
+                TestCase {
+                    this: Collection::from_parts(
+                        BTreeMap::from([("bar", Kind::bytes())]),
+                        Kind::regex().or_null(),
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:232]: Often, the starting states are mostly there for me to mix-and-match different type states so that there aren't any weird bugs that creep in that happen only for one state, but not the other.

Small chance, and usually this would be better tested with property tests and the like, but it's a small price to pay, for a tiny bit of extra security.

File: lib/value/src/kind/collection.rs [hunk 18]
```diff
@@ -0,0 +1,403 @@
...
+                },
+            ),
+        ]) {
+            assert_eq!(this.is_superset(&other), want, "{}", title);
+        }
+    }
+
+    #[test]
+    #[allow(clippy::too_many_lines)]
+    fn test_merge() {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:287]: Good catch. I added these tests before I added that strategy (since it was needed to pass some VRL tests). I updated the test case by copy/pasting the same strategy into all cases, but forgot to add new cases testing the different strategies. Will do so.

File: lib/value/src/kind/collection/exact.rs [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+    pub(super) regex: bool,
+    pub(super) null: bool,
+
+    // We don't need to support nested objects, because `Exact` is only used for "unknown" fields.
+    // So it only applies to a single level. If we wanted to set a nested field, we would have
+    // a "known" field (e.g. `foo`) and then have that contain either `Unknown::Any` or
+    // `Unknown::Exact`.
+    pub(super) object: bool,
+    pub(super) array: bool,
+}
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection/exact.rs:28]: I explicitly didn't want to use it because we use(d) it in the old implementation, and while it is cool and all at first, it just comes with an API that is broader than we need, and it caused more confusion than it was worth over the last year+.

The most frequently confusing example I can remember is that bitflags has an `all()` function to initialize the type, but we call it `any()`, as that better fits the mental model of progressive type checking, but every so often, that method popped up in the code because people's editors auto-completed it, and it _works_, but it's confusing to see it being used next to `any()`.

TL;DR: at a surface, it makes sense, but after having used it for more than a year, I found it to be more confusing than worthwhile.

File: lib/value/src/kind/collection/exact.rs [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+            null: true,
+            object: true,
+            array: true,
+        }
+    }
+
+    /// Check if `self` is a superset of `other`.
+    ///
+    /// Meaning, if `other` has a type set to `true`, then `self` needs to as well.
+    pub fn is_superset(&self, other: &Self) -> bool {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection/exact.rs:48]: Correct, but as explained above, I explicitly avoided the crate due to my experience with it in the "old" `Kind` implementation.

File: lib/value/src/kind/collection/index.rs [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+    }
+}
+
+impl<T: Into<Index>> std::ops::Sub<T> for Index {
+    type Output = Self;
+
+    fn sub(self, other: T) -> Self {
+        Self(self.0 - other.into().0)
+    }
+}
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection/index.rs:43]: I believe in `remove_at_path`, when down-shifting indices.

File: lib/value/src/kind/collection/unknown.rs [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+    #[must_use]
+    pub fn is_exact(&self) -> bool {
+        matches!(self, Self::Exact(_))
+    }
+
+    /// Check if `self` is a superset of `other`.
+    ///
+    /// Meaning, if `self` is `Any`, then it's always a superset of `other`, otherwise its
+    /// accumulative types need to be a superset of `other`.
+    pub fn is_superset(&self, other: &Self) -> bool {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection/unknown.rs:54]: I'm not certain why Clippy doesn't warn in some cases, but will add.

File: lib/value/src/kind/collection/unknown.rs [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+            (Self::Any, _) | (_, Self::Any) => Self::Any,
+            (Self::Exact(lhs), Self::Exact(rhs)) => Self::Exact(lhs | rhs),
+        }
+    }
+}
+
+impl BitXor for Unknown {
+    type Output = Self;
+
+    /// If the state of `Unknown` fields is `any`, this method always returns `any`.
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection/unknown.rs:83]: It should, I didn't pay much attention to these impls, other than having them for consistency, but it's actually one of the other regrets I have of the old `Kind` impl, and I've since removed these from the follow-up PR. It's just too confusing in everyday use, given the complexities that come from the conceptual model of progressive type checking.

File: lib/value/src/kind/comparison.rs (hunk 4 above)
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:175]: `is_*` returns a boolean, not sure how that would work?

File: lib/value/src/kind/comparison.rs (hunk 6 above)
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs:270]: In VRL, you can search the existing `type_def` in master, or wait for the follow-up PR.

File: lib/value/src/kind/collection/exact.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection/exact.rs:28]: I'll add that we _could_ wrap the `bitflags` type in a newtype, but ultimately, I just don't think the juice is worth the squeeze.

File: lib/value/src/kind/merge.rs [hunk 24]
```diff
@@ -0,0 +1,408 @@
...
+    }
+
+    /// Check if `append` strategy is enabled.
+    #[must_use]
+    pub fn is_append(self) -> bool {
+        matches!(self, Self::Append)
+    }
+}
+
+impl std::ops::BitOr for Kind {
```
Comment (spencergilbert) [inline lib/value/src/kind/merge.rs:125]: I think elsewhere in the PR you imported this at the top, is there a reason why they weren't all done the same?

File: lib/value/src/kind/insert.rs [hunk 25]
```diff
@@ -0,0 +1,661 @@
...
+
+    /// Remove the existing `Kind` states, replacing it for a singular object or array
+    /// `Kind` state.
+    Replace,
+
+    /// Reject the insertion, returning an error.
+    Reject,
+}
+
+/// The strategy to use when the leaf segment already has a `Kind` present.
```
Comment (spencergilbert) [inline lib/value/src/kind/insert.rs:31]: Leaf segment == final path segment?

Comment (JeanMertz) [reply lib/value/src/kind/insert.rs:31]: Correct. It's common terminology for tree-like structures, although paths aren't necessarily a tree structure, so if you think it's confusing, let me know and we can change it.

File: lib/value/src/kind/merge.rs (hunk 24 above)
Comment (JeanMertz) [reply lib/value/src/kind/merge.rs:125]: No, will check for consistency.

File: lib/value/src/kind/accessor.rs [hunk 26]
```diff
@@ -0,0 +1,543 @@
...
+                TestCase {
+                    kind: Kind::object(BTreeMap::from([("foo".into(), Kind::integer())])),
+                    path: "bar".into(),
+                    want: None,
+                },
+            ),
+            (
+                "array w/ matching path",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:255]: It's not, it's just a random number I picked for testing (arrays can have gaps in their indices, if we don't have any information for a given position (this is what `unknown` details are for in a collection).

File: lib/value/src/kind/accessor.rs [hunk 27]
```diff
@@ -0,0 +1,543 @@
...
+    /// Nest the given [`Kind`] into a provided path.
+    ///
+    /// For example, given an `integer` kind and a path `.foo`, a new `Kind` is returned that is
+    /// known to be an object, of which the `foo` field is known to be an `integer`.
+    #[must_use]
+    pub fn nest_at_path(mut self, path: &Lookup<'_>) -> Self {
+        fn object_from_field(field: &lookup::Field<'_>, kind: Kind) -> Kind {
+            let map = BTreeMap::from([(field.into(), kind)]);
+            Kind::object(map)
+        }
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:60]: It is. It's a helper function to allow recursion down below.

File: lib/value/src/kind/collection.rs [hunk 28]
```diff
@@ -0,0 +1,403 @@
...
+    }
+
+    /// Get a mutable reference to "known" field value kinds.
+    pub fn known_mut(&mut self) -> &mut BTreeMap<T, Kind> {
+        &mut self.known
+    }
+
+    /// Get the "unknown" field value kind.
+    #[must_use]
+    pub fn as_unknown(&self) -> &Unknown {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/collection.rs:93]: Yeah, exactly, there's already an `unknown` for initializing. However, I've updated parts of this code in a follow-up PR that's coming today, where I have `from_unknown` to do the initialization, and then there's `unknown` and `unknown_mut` similar to `known` and `known_mut`.

File: lib/value/src/kind/accessor.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply lib/value/src/kind/accessor.rs:60]: Interesting, I don't know if I remember seeing a `fn` declared inside of another `fn` - thanks!

File: lib/value/src/kind/insert.rs (hunk 25 above)
Comment (spencergilbert) [reply lib/value/src/kind/insert.rs:31]: Nope - leaf is fine, just wanted to double check as I mid brain melt at this point :)

File: lib/value/src/kind/comparison.rs (hunk 4 above)
Comment (StephenWakely) [reply lib/value/src/kind/comparison.rs:175]: Perhaps: 

```rust
let a = |b| if b { 1 } else { 0 };

a(self.is_bytes()) + 
a(self.is_integer()) + 
a(self.is_float()) + 
a(self.is_boolean()) + 
a(self.is_timestamp()) + 
a(self.is_regex()) + 
a(self.is_null()) + 
a(self.is_array()) + 
a(self.is_object()) == 1
```

**Comment by fuchsnj** [comment]
I think `"lib/value"` should be added to the workspace members list in the root `Cargo.toml`

File: lib/value/Cargo.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/Cargo.toml:4]: Actually, it appears we haven't updated this in other libraries either. I think it's best to delegate this to a separate PR. I created an issue to track this at #11103.

File: lib/value/src/kind/comparison.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/value/src/kind/comparison.rs]: After the latest refactor, this code is gone. I haven't added any more doc-tests, but we could do so in the future if we think it adds value.

**Review by pablosichert — APPROVED** [review]
Awesome work, looks good to me.

File: lib/value/src/kind/accessor.rs [hunk 29]
```diff
@@ -0,0 +1,677 @@
...
+impl Kind {
+    /// Find the [`Kind`] at the given path.
+    ///
+    /// If the path points to root, then `self` is returned, otherwise `None` is returned if `Kind`
+    /// isn't an object or array. If the path points to a non-existing element in an existing collection,
+    /// then the collection's `unknown` `Kind` variant is returned.
+    ///
+    /// Negative indexing always returns `None`, as the type system doesn't know the exact size of
+    /// an array, and thus cannot count backward.
+    #[must_use]
```
Comment (StephenWakely) [inline lib/value/src/kind/accessor.rs:16]: I'm not convinced these `must_use` annotations are useful since I can't see any scenario where someone would call this function and forget to use the result. 

I think the intention behind this is for situations where the functions success path returns no useful result, but the error path returns an error. A developer could forget to handle the error and think their code is working fine.

Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs:16]: The documentation (and RFC) has examples that specifically call out any use-case where the result of a function should be used for that function to be useable (e.g. for a pure function that has no side effects).

See: https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute

In fact, there's [ongoing work](https://github.com/rust-lang/rust/issues/89692) to expand this attribute across Rust's stdlib as well (such as `len()` and `is_empty()`).

See also this "[for reviewers](https://std-dev-guide.rust-lang.org/code-considerations/design/must-use.html#for-reviewers)" section, which calls out:

> Look for any legitimate use-cases where `#[must_use]` will cause callers to explicitly ignore values. If these are common then `#[must_use]` probably isn't appropriate.

In this (and other) case(s), calling `find_at_path`, but _not_ using the return value is clearly a programmer bug, and so a valid case to apply `#[must_use]` to.

Personally, I don't mind _too much_ whether we add it or not, but it's a nice tool to have in our toolbox to prevent any bugs in API misuse down the line, and while it does add a bit of clutter, I'm mostly on the side that the clutter is worth the extra safety.

File: lib/value/src/kind/accessor.rs (hunk 3 above)
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs]: I decided to reject coalesced nesting for now. We can revisit this later.

File: lib/value/src/kind/accessor.rs (hunk 5 above)
Comment (JeanMertz) [reply lib/value/src/kind/accessor.rs]: I've instead opted to reject coalesced path deletion (in this library) for now.
